codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/OperationInterceptors.java

```diff
@@ -42,71 +29,69 @@ public byte getOrder() {
-                        }, "OperationSerializer", middleware.Before)
-                    }
-
-                    func addInterceptAfterSerialization(stack *middleware.Stack, opts Options) error {
-                        return stack.Serialize.Insert(&smithyhttp.InterceptAfterSerialization{
-                            Interceptors: opts.Interceptors.AfterSerialization,
-                        }, "OperationSerializer", middleware.After)
-                    }
-
-                    func addInterceptBeforeSigning(stack *middleware.Stack, opts Options) error {
-                        return stack.Finalize.Insert(&smithyhttp.InterceptBeforeSigning{
-                            Interceptors: opts.Interceptors.BeforeSigning,
-                        }, "Signing", middleware.Before)
-                    }
-
-                    func addInterceptAfterSigning(stack *middleware.Stack, opts Options) error {
-                        return stack.Finalize.Insert(&smithyhttp.InterceptAfterSigning{
-                            Interceptors: opts.Interceptors.AfterSigning,
-                        }, "Signing", middleware.After)
-                    }
-
-                    func addInterceptTransmit(stack *middleware.Stack, opts Options) error {
-                        return stack.Deserialize.Add(&smithyhttp.InterceptTransmit{
-                            BeforeTransmit: opts.Interceptors.BeforeTransmit,
-                            AfterTransmit:  opts.Interceptors.AfterTransmit,
-                        }, middleware.After)
-                    }
-
-                    func addInterceptBeforeDeserialization(stack *middleware.Stack, opts Options) error {
-                        return stack.Deserialize.Insert(&smithyhttp.InterceptBeforeDeserialization{
-                            Interceptors: opts.Interceptors.BeforeDeserialization,
-                        }, "OperationDeserializer", middleware.After) // (deserialize stack is called in reverse)
-                    }
+                    func addInterceptors(stack *middleware.Stack, opts Options) error {
+                        // middlewares are expensive, don't add all of these interceptor ones unless the caller
+                        // actually has at least one interceptor configured
+                        //
+                        // at the moment it's all-or-nothing because some of the middlewares here are responsible for
+                        // setting fields in the interceptor context for future ones
+                        if len(opts.Interceptors.BeforeExecution) == 0 &&
+                            len(opts.Interceptors.BeforeSerialization) == 0 && len(opts.Interceptors.AfterSerialization) == 0 &&
+                            len(opts.Interceptors.BeforeRetryLoop) == 0 &&
+                            len(opts.Interceptors.BeforeAttempt) == 0 &&
+                            len(opts.Interceptors.BeforeSigning) == 0 && len(opts.Interceptors.AfterSigning) == 0 &&
```

Comment:
very small nit on ordering, but this seems like it should go after all the `BeforeX`, or we could put them in the order they are added below on `errors.Join`